### DISCUSSION TITLE
Observer not working when passing ObservableList to child widget

### DISCUSSION THREAD

**Discussion by vytautas-pranskunas-** [mobxjs/mobx.dart#765] [General]
Hello,

recently we have discovered a bug where events were not showing on calendar date change. This bug was caused by the fact that reactivity were not triggered on filteredDates ObservableList change because it was passed down to child widget. However when I have added observable list usage at top level it started to acct as expected.

Not triggered
```
Observer(builder: (_) {
        return EventsDisplay(_store.filteredEvent);
      }),
```

Triggered:
```
Observer(builder: (_) {
        return _store.filteredEvent.isEmpty ? Container() : EventsDisplay(_store.filteredEvent);
      }),
```

Any ideas why?

**Comment by vytautas-pranskunas-** [comment]
Did not get it - why it is discussion if thats a bug? :D

**Reply by pavanpodila** [reply]
Is `filteredEvent` an observable ? If not, the Observer has no way of knowing when to react

**Reply by vytautas-pranskunas-** [reply]
As i mentioned - it is. It is clearly visible in example that first one is not working but when I am using it in parent it works.

**Reply by fzyzcjy** [reply]
Could you please share the code of `_store`?

**Reply by vytautas-pranskunas-** [reply]
@observable
ObservableList<Event> filteredEvent = ObservableList<Event>();

**Reply by fzyzcjy** [reply]
That sounds quite reasonable. Could you please share the *full* code of the class containing filteredEvent? And moreover, the `.g.dart` file. For example, if the code generator does not run for some reason, the generated code is stale, then that `@observable` does not take effect

**Reply by vytautas-pranskunas-** [reply]
sure:

https://drive.google.com/drive/folders/10KwdIg6V_SgaidqrEINe0G2dYRpR6hKs?usp=sharing

code generator runs normally because I would not release an app :)

**Reply by fzyzcjy** [reply]
Great. Btw next time you can use gist.github.com

**Reply by fzyzcjy** [reply]
`return EventsDisplay(_store.filteredEvent);` will only make it rebuild when the filteredEvent is re-assigned. On the other hand, if content within it changes and you have stored filteredEvent variable somewhere then it will not trigger rebuild.

Anyway, if you think this is a bug, please make a "minimal reproducible sample" :)

**Reply by vytautas-pranskunas-** [reply]
I need to trigger a rebuild thats a thing
I understand that @observable decorator triggers when reference changes but ObservableList should be triggered when content changes! And what i was expecting that i can pass ObservableList down to child and used Observer(builder: () => child) hoping that child internals gets re-rendered when I am adding more events.

**Reply by subzero911** [reply]
hey @vytautas-pranskunas- look at my topic https://github.com/mobxjs/mobx.dart/issues/851#issuecomment-1242895544
I encountered the same issue, and we found a solution

**Reply by vytautas-pranskunas-** [reply]
@subzero911 thanks - that makes sense. Anyway it shold be added to docs :)

**Comment by deadsoul44** [comment]
Isn’t it due to the fact that observers are able to react to the observables only inside their immediate context? EventsDisplay widget has its own context.

**Reply by fzyzcjy** [reply]
Agree. `@observable ObservableList` is needed in this case. Not sure whether op misuse `ObservableList` instead of `@observable ObservableList`

**Reply by vytautas-pranskunas-** [reply]
it is @observable of @observableList
```
@observable
  ObservableList<Event> filteredEvent = ObservableList<Event>();
```

**Reply by subzero911** [reply]
> EventsDisplay widget has its own context.

What if it's a widget whose context I can't access?
In my example below, I use YandexMap, which is from the third-party package. But I need to rebuild it somehow, in response to ObservableList changes.

**Comment by subzero911** [comment]
I encountered the same problem: https://github.com/mobxjs/mobx.dart/issues/851
I pass ObservableList into YandexMap (it's a side package I can't edit) and mutate it with `list.add(element)`.
Observer tells in a console that it just doesn't see any observables inside!
But if I pass `Observable<List>` and change the list like `list.value = [...newList]`, Observer rebuilds it.
I consider it to be a bug. Observer should be aware of ObservableList inside.
